### PR TITLE
Add CDC partitioner (Version 4.9.0)

### PR DESCRIPTION
**Pull request by avelanarius**
Add new class `CDCToken` which implements a Scylla CDC partitioner, as defined here:

https://github.com/scylladb/scylla/blob/master/cdc/cdc_partitioner.cc

Add new `getPartitioner()` method in `Request` and `PreparedStatement` in order to be able to use custom partitioner on CDC log tables (and potentially other in the future).

Modify `Conversions.java` to set CDCPartitioner for CDC log prepared statements.

Change `TokenMap`, `CqlRequestHandler` to be able to use custom partitioner (or cluster-wide if null is provided).

A new `TokenLong64` type is introduced to allow comparisons between `Murmur3Token` and `CDCToken` (both backed by 64-bit integers), as an existing `TokenMap`(with `Murmur3Token`) is used.

Similarly to https://github.com/scylladb/java-driver/pull/35, I have performed manual testing (running queries with tracing and checking if they are sent to proper node and shard).

### CODE REVIEW

**Comment by avelanarius** [comment]
@haaawk ping

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/CDCToken.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package com.datastax.oss.driver.internal.core.metadata.token;
+
+import net.jcip.annotations.Immutable;
```
Comment (haaawk) [inline core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/CDCToken.java:3]: I'm not a great fan of nonstandard annotations like this. Could we please remove it?

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/CDCTokenFactory.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      logger.warn(
+          "CDC partition key has invalid length: expected {} bytes, but got {} bytes",
+          CDC_PARTITION_KEY_LENGTH,
+          length);
+    }
+    if (length < 8) {
+      return MIN_TOKEN;
+    }
+
+    long upperDword = partitionKey.getLong(offset + 0);
```
Comment (haaawk) [inline core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/CDCTokenFactory.java:45]: offset + 0?

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/CDCTokenFactory.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+import org.slf4j.LoggerFactory;
+
+@ThreadSafe
+public class CDCTokenFactory implements TokenFactory {
+
+  public static final String PARTITIONER_NAME = "com.scylladb.dht.CDCPartitioner";
+
+  public static final CDCToken MIN_TOKEN = new CDCToken(Long.MIN_VALUE);
+  public static final CDCToken MAX_TOKEN = new CDCToken(Long.MAX_VALUE);
+
```
Comment (haaawk) [inline core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/CDCTokenFactory.java:18]: Why are those fields public?

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/ByteOrderedTokenFactory.java [hunk 4]
```diff
@@ -16,6 +16,7 @@
 package com.datastax.oss.driver.internal.core.metadata.token;
 
 import com.datastax.oss.driver.api.core.metadata.token.Token;
+import com.datastax.oss.driver.api.core.metadata.token.TokenFactory;
```
Comment (haaawk) [inline core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/ByteOrderedTokenFactory.java]: Why is this needed when you haven't changed this class?

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/RandomTokenFactory.java [hunk 5]
```diff
@@ -16,6 +16,7 @@
 package com.datastax.oss.driver.internal.core.metadata.token;
 
 import com.datastax.oss.driver.api.core.metadata.token.Token;
+import com.datastax.oss.driver.api.core.metadata.token.TokenFactory;
```
Comment (haaawk) [inline core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/RandomTokenFactory.java]: Why is this needed when you haven't changed this class?

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/Murmur3TokenFactory.java [hunk 6]
```diff
@@ -16,6 +16,7 @@
 package com.datastax.oss.driver.internal.core.metadata.token;
 
 import com.datastax.oss.driver.api.core.metadata.token.Token;
+import com.datastax.oss.driver.api.core.metadata.token.TokenFactory;
```
Comment (haaawk) [inline core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/Murmur3TokenFactory.java]: Why is this needed when you haven't changed this class?

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/Murmur3Token.java [hunk 7]
```diff
@@ -15,49 +15,14 @@
...
-import com.datastax.oss.driver.shaded.guava.common.primitives.Longs;
-import edu.umd.cs.findbugs.annotations.NonNull;
 import net.jcip.annotations.Immutable;
 
 /** A token generated by {@code Murmur3Partitioner}. */
 @Immutable
-public class Murmur3Token implements Token {
-
-  private final long value;
+public class Murmur3Token extends TokenLong64 {
```
Comment (haaawk) [inline core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/Murmur3Token.java:22]: This extraction should be in a separate commit for the sake of easier review.

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/TokenFactoryRegistry.java [hunk 8]
```diff
@@ -15,6 +15,8 @@
  */
 package com.datastax.oss.driver.internal.core.metadata.token;
 
+import com.datastax.oss.driver.api.core.metadata.token.TokenFactory;
```
Comment (haaawk) [inline core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/TokenFactoryRegistry.java]: Why is this needed when you haven't changed this class?

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/CDCTokenFactory.java (hunk 3 above)
Comment (avelanarius) [reply core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/CDCTokenFactory.java:18]: No good reason aside from the fact that other (already implemented) TokenFactories also have them public:

https://github.com/scylladb/java-driver/blob/017b27315aabbc813bfda4070f778cbdb37907ec/core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/Murmur3TokenFactory.java#L29-L30

(consistency)

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/CDCTokenRange.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    if (range.compareTo(BigInteger.ZERO) < 0) {
+      range = range.add(RING_LENGTH);
+    }
+
+    List<BigInteger> values = super.split(start, range, RING_END, RING_LENGTH, numberOfSplits);
+    List<Token> tokens = Lists.newArrayListWithExpectedSize(values.size());
+    for (BigInteger value : values) {
+      tokens.add(new CDCToken(value.longValue()));
+    }
+    return tokens;
```
Comment (haaawk) [inline core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/CDCTokenRange.java]: You could replace this with `values.stream().map(v -> new CDCToken(v.longValue())).collect(Collectors.<Token>toList())` or `values.stream().mapToLong(v -> v.longValue()).map(CDCToken::new).collect(Collectors.<Token>toList())`

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/DefaultTokenFactoryRegistry.java [hunk 10]
```diff
@@ -15,6 +15,7 @@
  */
 package com.datastax.oss.driver.internal.core.metadata.token;
 
+import com.datastax.oss.driver.api.core.metadata.token.TokenFactory;
```
Comment (haaawk) [inline core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/DefaultTokenFactoryRegistry.java]: Why is this needed when you haven't changed this class?

File: core/src/main/java/com/datastax/oss/driver/api/core/metadata/token/TokenFactory.java [hunk 11]
```diff
@@ -13,10 +13,8 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package com.datastax.oss.driver.internal.core.metadata.token;
+package com.datastax.oss.driver.api.core.metadata.token;
 
-import com.datastax.oss.driver.api.core.metadata.token.Token;
-import com.datastax.oss.driver.api.core.metadata.token.TokenRange;
 import java.nio.ByteBuffer;
```
Comment (haaawk) [inline core/src/main/java/com/datastax/oss/driver/api/core/metadata/token/TokenFactory.java]: Why?

File: core/src/main/java/com/datastax/oss/driver/api/core/metadata/TokenMap.java [hunk 12]
```diff
@@ -51,12 +52,19 @@
...
    * Builds a token from a partition key.
    *
+   * @param partitioner the partitioner to use or {@code null} for this TokenMap's partitioner.
    * @param partitionKey the partition key components, in their serialized form (which can be
    *     obtained with {@link TypeCodec#encode(Object, ProtocolVersion)}. Neither the individual
    *     components, nor the vararg array itself, can be {@code null}.
    */
   @NonNull
-  Token newToken(@NonNull ByteBuffer... partitionKey);
+  Token newToken(TokenFactory partitioner, @NonNull ByteBuffer... partitionKey);
```
Comment (haaawk) [inline core/src/main/java/com/datastax/oss/driver/api/core/metadata/TokenMap.java]: Maybe annotate partitioner as Nullable?

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/DefaultMetadata.java [hunk 13]
```diff
@@ -20,10 +20,10 @@
 import com.datastax.oss.driver.api.core.metadata.Node;
 import com.datastax.oss.driver.api.core.metadata.TokenMap;
 import com.datastax.oss.driver.api.core.metadata.schema.KeyspaceMetadata;
+import com.datastax.oss.driver.api.core.metadata.token.TokenFactory;
```
Comment (haaawk) [inline core/src/main/java/com/datastax/oss/driver/internal/core/metadata/DefaultMetadata.java]: Please don't do changes like this. They will create merge conflicts later and give us absolutely no value.

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/FullNodeListRefresh.java [hunk 14]
```diff
@@ -16,9 +16,9 @@
 package com.datastax.oss.driver.internal.core.metadata;
 
 import com.datastax.oss.driver.api.core.metadata.Node;
+import com.datastax.oss.driver.api.core.metadata.token.TokenFactory;
```
Comment (haaawk) [inline core/src/main/java/com/datastax/oss/driver/internal/core/metadata/FullNodeListRefresh.java]: Please revert.

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/InitialNodeListRefresh.java [hunk 15]
```diff
@@ -17,8 +17,8 @@
 
 import com.datastax.oss.driver.api.core.metadata.EndPoint;
 import com.datastax.oss.driver.api.core.metadata.Node;
+import com.datastax.oss.driver.api.core.metadata.token.TokenFactory;
```
Comment (haaawk) [inline core/src/main/java/com/datastax/oss/driver/internal/core/metadata/InitialNodeListRefresh.java]: Please revert.

**Comment by haaawk** [comment]
The change looks ok at the first sight but there's so much noise that it's hard to focus on the important bits.
Please remove all the changes that are not relevant to the addition of CDC partitioner. Namely:
1. Reformating code that hasn't changed
2. Reordering imports
3. Changing imports even though the class hasn't been touched.
4. Anything else that's not strictly needed for CDC partitioner support.

Also please do Murmur3Token -> Long64Token refactoring as a separate commit.

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/ByteOrderedTokenFactory.java (hunk 4 above)
Comment (avelanarius) [reply core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/ByteOrderedTokenFactory.java]: The whole story:

1. I added `TokenFactory getPartitioner()` to `com.datastax.oss.driver.api.core.cql.PreparedStatement` (`api` NOT `internal` package).
2. After that, automatic checker in build system complained that `TokenFactory` is from `com.datastax.oss.driver.internal.core.metadata.token` (`internal`!). It complained that `api` cannot import stuff from `internal` (probably for automatic docs purposes). Correction: Probably not import but expose "to the external world" - in this case by making `internal` class as a return type.
3. That forced me to move `TokenFactory` from `internal` to `api`:
image
4. Other classes in `com.datastax.oss.driver.internal.core.metadata.token` used `TokenFactory`. Because before the change it was in the same package, no import was needed. After moving `TokenFactory` from `internal` to `api`, this import is needed.

Comment (haaawk) [reply core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/ByteOrderedTokenFactory.java]: Ok makes sense but why isn't it a separate commit which describes what's going on and why?

**Comment by avelanarius** [comment]
@haaawk Split commit into multiple smaller commits. 

Regarding points 1-4: All those imports are related mostly to moving `TokenFactory` from `internal` to `api`. When I was splitting the commits, I went practically over the every line once again and verified that this is the case.

Also, as to reformatting and reordering imports - it's not my mistake or some IDE left-over: the style checker in build system **very strictly** enforces this - including reordering imports and deleting old (unused) ones.

File: core/src/main/java/com/datastax/oss/driver/api/core/metadata/TokenMap.java (hunk 12 above)
Comment (avelanarius) [reply core/src/main/java/com/datastax/oss/driver/api/core/metadata/TokenMap.java]: Done in amended commits.

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/CDCTokenRange.java (hunk 9 above)
Comment (avelanarius) [reply core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/CDCTokenRange.java]: Fixed in amended commit.

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/Murmur3Token.java (hunk 7 above)
Comment (avelanarius) [reply core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/Murmur3Token.java:22]: Separated into separate commit.

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/CDCTokenFactory.java (hunk 2 above)
Comment (avelanarius) [reply core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/CDCTokenFactory.java:45]: To indicate that `offset` is a starting offset and we will read first long (`upperDword`) and second long (`lowerDword`), explicitly stating the "internal" offsets of those two values.

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/ByteOrderedTokenFactory.java (hunk 4 above)
Comment (avelanarius) [reply core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/ByteOrderedTokenFactory.java]: Separated into a52dc57efcd1ff7e1026ec0578081b8e175c9a1c, with description in the commit message.

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/CDCTokenFactory.java (hunk 3 above)
Comment (haaawk) [reply core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/CDCTokenFactory.java:18]: ok.

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/CDCTokenFactory.java (hunk 2 above)
Comment (haaawk) [reply core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/CDCTokenFactory.java:45]: I don't see much value in it but ok.

**Review by haaawk — CHANGES_REQUESTED** [review]
Thanks @avelanarius it's much easier to review now.
The change looks good but I would like to suggest some improvements:
1. Don't move TokenFactory class to api - this is a very intrusive change which I expect to cause a lot of pain in the future while we will be merging with upstream. Please instead define a new Partition interface that's passed around and has the hash method. Internally implemented using TokenFactory but without exposing it.
2.  Please extract partitioner method from Conversions.java into helper class and use this helper class inside Conversions.java. Just to minimize the risk of merge conflicts.

Otherwise the change LGTM

**Comment by avelanarius** [comment]
@haaawk Regarding the first point, is this a good way to do it?:

`TokenFactory` is already an interface, so change it to abstract class which will `implements` new `Partition` interface. This `TokenFactory` change will necessitate changing `implements` to `extends` in all token factories classes (a few files). `Partition` interface will consist of all methods of `TokenFactory`. 

To limit the lines of code/file count changed, the only place that will use `Partition` will be `getPartition()`. In all other places `getPartition()` will immediately be cast to `TokenFactory` (other way would require changing `TokenFactory` to `Partition` in large number of places).

**Comment by avelanarius** [comment]
@haaawk Pushed amended commits - fixed those two points you mentioned.

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/TokenFactory.java [hunk 16]
```diff
@@ -15,26 +15,23 @@
  */
 package com.datastax.oss.driver.internal.core.metadata.token;
 
+import com.datastax.oss.driver.api.core.metadata.token.Partitioner;
 import com.datastax.oss.driver.api.core.metadata.token.Token;
 import com.datastax.oss.driver.api.core.metadata.token.TokenRange;
-import java.nio.ByteBuffer;
 
 /** Manages token instances for a partitioner implementation. */
-public interface TokenFactory {
```
Comment (haaawk) [inline core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/TokenFactory.java:23]: I don't understand why this change is needed. Why can't TokenFactory by an interface that extends Partitioner interface?

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/TokenFactory.java [hunk 17]
```diff
@@ -15,26 +15,23 @@
  */
 package com.datastax.oss.driver.internal.core.metadata.token;
 
+import com.datastax.oss.driver.api.core.metadata.token.Partitioner;
 import com.datastax.oss.driver.api.core.metadata.token.Token;
 import com.datastax.oss.driver.api.core.metadata.token.TokenRange;
-import java.nio.ByteBuffer;
```
Comment (haaawk) [inline core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/TokenFactory.java:20]: Please don't do unrelated cleanups.

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/TokenFactory.java (hunk 16 above)
Comment (haaawk) [reply core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/TokenFactory.java:23]: Even better would be for TokenFactory.getPartitionerName to be converted into TokenFactory.getPartitioner that returns Partitioner. I'm not sure whether it would fit well with the code though.

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/DefaultTokenMap.java [hunk 18]
```diff
@@ -171,9 +176,14 @@ public TokenRange newTokenRange(@NonNull Token start, @NonNull Token end) {
...
-  public Set<Node> getReplicas(@NonNull CqlIdentifier keyspace, @NonNull ByteBuffer partitionKey) {
+  public Set<Node> getReplicas(
+      @NonNull CqlIdentifier keyspace,
+      @Nullable Partitioner partitioner,
+      @NonNull ByteBuffer partitionKey) {
     KeyspaceTokenMap keyspaceMap = getKeyspaceMap(keyspace);
-    return (keyspaceMap == null) ? Collections.emptySet() : keyspaceMap.getReplicas(partitionKey);
+    return (keyspaceMap == null)
+        ? Collections.emptySet()
+        : keyspaceMap.getReplicas((TokenFactory) partitioner, partitionKey);
```
Comment (haaawk) [inline core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/DefaultTokenMap.java]: Please add a comment explaining why it is safe to cast to TokenFactory here.

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/KeyspaceTokenMap.java [hunk 19]
```diff
@@ -96,8 +96,11 @@ private KeyspaceTokenMap(
     return tokenRangesByNode.get(replica);
   }
 
-  Set<Node> getReplicas(ByteBuffer partitionKey) {
-    return getReplicas(tokenFactory.hash(partitionKey));
+  Set<Node> getReplicas(TokenFactory partitioner, ByteBuffer partitionKey) {
```
Comment (haaawk) [inline core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/KeyspaceTokenMap.java]: Why not to take Partitioner instead of TokenFactory here and remove casting from DefaultTokenMap.getReplicas?

**Comment by haaawk** [comment]
LGTM in general but I don't understand why TokenFactory can't be an interface anymore.
Also, I think it might be possible to remove casting around getReplicas.

File: core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/TokenFactory.java (hunk 17 above)
Comment (avelanarius) [reply core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/TokenFactory.java:20]: Again, this is caused by automatic style checker:
```
[ERROR] Found 1 non-complying files, failing build
[ERROR] To fix formatting errors, run "mvn com.coveo:fmt-maven-plugin:format"
```
(after adding back this import)

Comment (haaawk) [reply core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/TokenFactory.java:20]: ok.

**Comment by avelanarius** [comment]
Changed `TokenFactory` back to `interface` (previously was under misunderstanding that interfaces cannot extend other interfaces...). Removed casting around `getReplicas`.

File: core/src/main/java/com/datastax/oss/driver/api/core/metadata/TokenMap.java [hunk 20]
```diff
@@ -99,15 +108,39 @@
...
+   * Shortcut for {@link #getReplicas(CqlIdentifier, Partitioner, ByteBuffer)
+   * getReplicas(CqlIdentifier.fromCql(keyspaceName), partitioner, partitionKey)}.
+   */
+  @NonNull
+  default Set<Node> getReplicas(
+      @NonNull String keyspaceName,
+      @Nullable Partitioner partitioner,
+      @NonNull ByteBuffer partitionKey) {
+    return getReplicas(CqlIdentifier.fromCql(keyspaceName), partitioner, partitionKey);
+  }
```
Comment (haaawk) [inline core/src/main/java/com/datastax/oss/driver/api/core/metadata/TokenMap.java:135]: since we've got those shortcuts, are really all those changes in the tests necessary?

Comment (avelanarius) [reply core/src/main/java/com/datastax/oss/driver/api/core/metadata/TokenMap.java:135]: Yes, I think so. I think that those tests directly call `getReplicas(keyspace, partitioner, partitionKey)` with `partitioner = null`, so mocks had to be updated. Without those shortcuts even more places would have to be fixed.

File: core/src/test/java/com/datastax/oss/driver/internal/core/loadbalancing/BasicLoadBalancingPolicyQueryPlanTest.java [hunk 21]
```diff
@@ -90,7 +91,9 @@ public void should_use_round_robin_when_no_request() {
 
     // Then
     then(tokenMap).should(never()).getReplicas(any(CqlIdentifier.class), any(Token.class));
-    then(tokenMap).should(never()).getReplicas(any(CqlIdentifier.class), any(ByteBuffer.class));
+    then(tokenMap)
+        .should(never())
+        .getReplicas(any(CqlIdentifier.class), isNull(), any(ByteBuffer.class));
```
Comment (haaawk) [inline core/src/test/java/com/datastax/oss/driver/internal/core/loadbalancing/BasicLoadBalancingPolicyQueryPlanTest.java:96]: Shouldn't isNull() be any()?

Comment (haaawk) [reply core/src/test/java/com/datastax/oss/driver/internal/core/loadbalancing/BasicLoadBalancingPolicyQueryPlanTest.java:96]: And in other should(never()) places.
